### PR TITLE
Automate choice of some params in BBH pipeline

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

Some of the initial parameters in our input file depend on the parameters used in the ID solve, and it'd be very tedious to try and update these manually in our pipeline. This PR adds logic to determine some constraint damping parameters and control system parameters automatically based on the masses, mass ratio, spins, and separation of the binary. These formulas were taken from `DoMultipleRuns.input` in SpEC.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

This PR should be merged after #5953

### CODE REVIEW

File: support/Pipelines/Bbh/Inspiral.py
Comment (nilsvu) [inline support/Pipelines/Bbh/Inspiral.py]: How about this for structuring the code better than SpEC: You put each of these sections into their own function that returns a dict. Then you call them like this:

```py
params = {
  # ...
}
params.update(
  control_system_parameters(total_mass=total_mass, ...)
)
return params
```

File: support/Pipelines/Bbh/Inspiral.yaml
Comment (nilsvu) [inline support/Pipelines/Bbh/Inspiral.yaml:180]: Note that you _can_ do simple math in these templates, e.g. `Constant: {{ 0.001 / (m_A + m_B) }}`. Maybe that's better than doing it in Python and passing it down? It would certainly make easy expressions a lot less verbose, and probably easier to understand.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: support/Pipelines/Bbh/Inspiral.yaml
Comment (knelli2) [reply support/Pipelines/Bbh/Inspiral.yaml:180]: I saw we already did that for some other parameters in the input file and I didn't like it because it's basically a python variable in a yaml file. And if we have math in the yaml like that it's a bit confusing. I'd prefer to do all the math in the python script and just treat the jinja templates as a find/replace with values. I think that's clearer for users if they ever look at these files (and clearer for me). Only one place to look for math rather than scattered through an input file

**Comment by knelli2** [comment]
Rebased and squashed in the suggestion about factoring out functions

File: support/Pipelines/Bbh/Inspiral.py
Comment (nilsvu) [inline support/Pipelines/Bbh/Inspiral.py]: (nitpick) For consistency with the code below I'd do this syntax, but that's just personal preference:

```py
return {
  "MaxDampingTimescale": max_damping_timescale,
  "KinematicTimescale": damping_time_base * total_mass,
  # ...
}
```

Have to precompute values used multiple times, but maybe that's better anyway.

File: support/Pipelines/Bbh/Inspiral.py
Comment (nilsvu) [inline support/Pipelines/Bbh/Inspiral.py]: (nitpick) don't need the `.keys()`

File: support/Pipelines/Bbh/Inspiral.py
Comment (nilsvu) [inline support/Pipelines/Bbh/Inspiral.py]: `supported_solutions`?

File: support/Pipelines/Bbh/Inspiral.py
Comment (nilsvu) [inline support/Pipelines/Bbh/Inspiral.py]: (nitpick) can be just `center_right - center_left`

File: support/Pipelines/Bbh/Inspiral.yaml
Comment (nilsvu) [reply support/Pipelines/Bbh/Inspiral.yaml:180]: 🤷‍♂️ don't know we can try it this way. I don't think doing easy math in the input file template is confusing, and it would reduce the number of parameters being computed and passed to the template. But I'm fine with doing it either way for now.

File: support/Pipelines/Bbh/Inspiral.py
Comment (nilsvu) [inline support/Pipelines/Bbh/Inspiral.py:30]: (more nitpicks) use `if else` here. In Python the scoping is more relaxed than in C++, so you don't need to define the variables before the `if`.

**Review by nilsvu — DISMISSED** [review]
you can squash

**Review by nilsvu — DISMISSED** [review]

**Comment by nilsvu** [comment]
@markscheel @geoffrey4444 could one of you take a look at this as well?

File: support/Pipelines/Bbh/Inspiral.py
Comment (markscheel) [inline support/Pipelines/Bbh/Inspiral.py]: Do we need these as parameters since they are functions of mass_left and mass_right?
(a potential bug is someone might pass in a `total_mass` that is not equal to `mass_left+mass_right`, and this relationship isn't tested in this function.)  If we really want all 4 parameters, then there should be `assert`s that they are correct.

File: support/Pipelines/Bbh/Inspiral.py
Comment (markscheel) [inline support/Pipelines/Bbh/Inspiral.py]: Similar; do we need total_mass?

File: support/Pipelines/Bbh/Inspiral.py
Comment (knelli2) [inline support/Pipelines/Bbh/Inspiral.py]: These masses are incorrect. They are the conformal masses for the Xcts solve. We actually need the Christodoulou/ADM masses for this. We can only get those from horizon finds after the ID solve.

I've been testing some runs with these (incorrect) masses and they don't matter too much for the control system params, but they are bad for the constraint damping params. My bbh runs die quickly. So I think we should hold off on this temporarily until we have horizon finds after the ID solve. Either I or @nilsvu probably need to do that.

File: support/Pipelines/Bbh/Inspiral.py
Comment (knelli2) [reply support/Pipelines/Bbh/Inspiral.py]: See my other comment. But yeah we should probably only take mass left/right and just recompute the total mass and mass ratio.

File: support/Pipelines/Bbh/Inspiral.py
Comment (nilsvu) [reply support/Pipelines/Bbh/Inspiral.py]: Yea I'll get you the horizon masses, though for equal-mass nonspinning you know they are ~0.5. Does the ~20% difference to the conformal masses matter so much for the constraint damping? Does the constraint damping work if you set the masses to 0.5?

Comment (knelli2) [reply support/Pipelines/Bbh/Inspiral.py]: Yes it matters quite a bit for the constraint damping parameters. Using the correct masses of 0.5 allowed the simulation to run much longer than with the incorrect masses.

File: support/Pipelines/Bbh/Inspiral.yaml
Comment (geoffrey4444) [reply support/Pipelines/Bbh/Inspiral.yaml:180]: As someone who in SpEC took every opportunity to do things in input files instead of code, I agree with Kyle: I've learned my lesson, and it's better to discourage users from doing math not in code, because it's easier for them to introduce mistakes, there are often fewer checks and safeguards, and it's more likely to end up with a yaml file not in version control than code

File: support/Pipelines/Bbh/Inspiral.py
Comment (geoffrey4444) [reply support/Pipelines/Bbh/Inspiral.py]: I agree with @knelli2 that you need to set the black-hole masses from the results of a horizon find, not the conformal masses

File: support/Pipelines/Bbh/Inspiral.py
Comment (geoffrey4444) [inline support/Pipelines/Bbh/Inspiral.py:103]: I'm not sure where you are using the spins, but these should be the true spins from a horizon find, not the bare, conformal spins

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]

File: support/Pipelines/Bbh/Inspiral.py
Comment (knelli2) [reply support/Pipelines/Bbh/Inspiral.py:103]: They are used for some control system parameters above. But yes I agree, this is included with using the correct masses from the horizon finds.

File: support/Pipelines/Bbh/Inspiral.py
Comment (nilsvu) [reply support/Pipelines/Bbh/Inspiral.py]: I have a branch to get horizon quantities from initial data here: https://github.com/nilsvu/spectre/commits/id_horizon_quantities/

You can call `find_horizon` on the initial data here in this function to get the quantities. We can also store these alongside the initial data later, maybe in the reductions file.

Comment (nilsvu) [reply support/Pipelines/Bbh/Inspiral.py]: Here's the PR to get the horizon masses and spins: https://github.com/sxs-collaboration/spectre/pull/6024

Comment (knelli2) [reply support/Pipelines/Bbh/Inspiral.py]: Once #6024 is merged, I'll use that here to get the correct masses and spins

Comment (nilsvu) [reply support/Pipelines/Bbh/Inspiral.py]: #6024 is merged now, so you can go ahead with this 👍

**Comment by knelli2** [comment]
@nilsvu Rebased and just combined changes since a decent amount changed. The tests will fail and I'm not really sure how to test this now though since it requires volume data to be written and be correct enough to find two horizons...

File: support/Pipelines/Bbh/Inspiral.py
Comment (nilsvu) [inline support/Pipelines/Bbh/Inspiral.py]: I moved this to a `postprocess_id` step that runs automatically after ID generation: #6082. Then the masses and spins are just stored alongside the ID in `BbhReductions.h5:Ah{A,B}.dat` and you can just read them here. That's better, right?

Comment (knelli2) [reply support/Pipelines/Bbh/Inspiral.py]: Naively, it does seem better. But we should probably discuss more offline about how this post-process step will interact with ID root find, ecc reduction, PB&J (which we'll need to do more planning about), and whether we want yet another step in the pipeline (too few and we don't have enough flexibility; too many and it gets confusing)

**Comment by knelli2** [comment]
Still need to test this in an actual pipeline

File: support/Pipelines/Bbh/Inspiral.yaml
Comment (nilsvu) [inline support/Pipelines/Bbh/Inspiral.yaml]: Have you tried this? I think this won't be wrapped in quotes, which could be a problem for paths that have spaces, and is definitely a problem in the list of subfile names below.

File: support/Python/Schedule.py
Comment (nilsvu) [inline support/Python/Schedule.py]: add a separate function for `np.float64` instead of these conditionals. The function is then added with `add_multi_representer` below.

File: support/Python/Schedule.py
Comment (nilsvu) [inline support/Python/Schedule.py]: why string? should probably be represented as a float in yaml, like Python's `float`

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: support/Pipelines/Bbh/Inspiral.yaml
Comment (nilsvu) [inline support/Pipelines/Bbh/Inspiral.yaml]: This was 0.2 before. I think that was important.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: support/Python/Schedule.py
Comment (knelli2) [reply support/Python/Schedule.py]: Because I tried it with just `float` and got an error

Comment (nilsvu) [reply support/Python/Schedule.py]: If you write it as a string it will read back as a string, which is wrong (and could be achieved simply by casting the float to a string before writing). I think the only acceptable behavior is if the `np.float64` reads back in as a `np.float64` or a `float`.

File: support/Pipelines/Bbh/Inspiral.yaml
Comment (knelli2) [reply support/Pipelines/Bbh/Inspiral.yaml]: Yeah this worked just fine. However, I didn't think about the spaces. I'll put these in quotes to account for that

**Comment by knelli2** [comment]
@nilsvu Ok I think this is ready now. The first commit has what worked for `np.float64`. I also added the control system and constraint damping params to when we are starting from SpEC ID because I realized I had forgot that and it'd fail without it. Note that from SpEC ID, the shape map still starts as a sphere, but this will change in a future PR once I have a lot more time dep option features added to the BCO domains.

**Review by nilsvu — APPROVED** [review]

**Pull request merged**
